crates/uv-auth/src/cache.rs

```diff
@@ -138,47 +74,198 @@ impl CredentialsCache {
         }
 
         // Insert an entry for requests including the username
-        if let Some(username) = credentials.username() {
-            self.insert_entry(
-                CredentialsCache::key(url, Some(username.to_string())),
-                credentials.clone(),
-            );
+        let username = credentials.to_username();
+        if username.is_some() {
+            let realm = (NetLoc::from(url), username.clone());
+            self.insert_realm(realm, credentials.clone());
         }
 
         // Insert an entry for requests with no username
-        self.insert_entry(CredentialsCache::key(url, None), credentials.clone());
+        self.insert_realm((NetLoc::from(url), Username::none()), credentials.clone());
+
+        // Insert an entry for the URL
+        let mut urls = self.urls.lock().unwrap();
+        urls.insert(url.clone(), credentials.clone());
     }
 
-    /// Private interface to update a cache entry.
-    fn insert_entry(&self, key: (NetLoc, Option<String>), credentials: Arc<Credentials>) -> bool {
+    /// Private interface to update a realm cache entry.
+    ///
+    /// Returns replaced credentials, if any.
+    fn insert_realm(
+        &self,
+        key: (NetLoc, Username),
+        credentials: Arc<Credentials>,
+    ) -> Option<Arc<Credentials>> {
         // Do not cache empty credentials
         if credentials.is_empty() {
-            return false;
+            return None;
         }
 
-        let mut store = self.store.lock().unwrap();
+        let mut realms = self.realms.lock().unwrap();
 
         // Always replace existing entries if we have a password
         if credentials.password().is_some() {
-            store.insert(key, credentials.clone());
-            return true;
+            return realms.insert(key, credentials.clone());
         }
 
         // If we only have a username, add a new entry or replace an existing entry if it doesn't have a password
-        let existing = store.get(&key);
+        let existing = realms.get(&key);
         if existing.is_none()
             || existing.is_some_and(|credentials| credentials.password().is_none())
         {
-            store.insert(key, credentials.clone());
-            return true;
+            return realms.insert(key, credentials.clone());
+        }
+
+        None
+    }
+}
+
+#[derive(Debug)]
+struct UrlTrie {
+    states: Vec<TrieState>,
+}
+
+#[derive(Debug, Default)]
+struct TrieState {
+    children: Vec<(String, usize)>,
+    value: Option<Arc<Credentials>>,
+}
```

Comment:
Thank you @BurntSushi!

I modified an implementation authored by Andrew to match prefixes with chunks of URL path segments rather than raw bytes. This prevents e.g. `example/foo1` from matching `example/foo`